### PR TITLE
ENH: Use IntergerArray to avoid forced conversion from integer to float

### PR DESCRIPTION
**Pull request by jiangyue12392**
- [ ] closes #xxxx
- [x] tests added / passed
- [ ] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This is a new attempt to solve issue #16918 after an attempt with PR #27073

### CODE REVIEW

**Comment by WillAyd** [comment]
Is there a way to make this work without adding a keyword to the DataFrame constructor? That approach seems off to me

Also be sure to add tests with this - its not entirely clear that this does anything as is

**Review by jreback — CHANGES_REQUESTED** [review]
@jiangyue12392 not what I wanted here. sure fixing maybe_convert_numeric to allow return integer_array  might close #27283 , but the rest is completely orthogonal and we cannot propagate a keyword thru DataFrame.

**Comment by jiangyue12392** [comment]
@WillAyd @jreback If the original implementation of using DataFrame to do the expansion of `list` of `dict` to 2d array and the type determination for columns is to be maintained then it seems to me that there is no way to circumvent propagating a new key in the DataFrame to make it backwards compatible. Is it more desirable to pull out the expansion and type determination logic from DataFrame and have some code duplications?

I will add tests later, but I want to know what is the more accepted methods here first.

**Comment by jreback** [comment]
@jiangyue12392 we are likely going to do the inference for expanding a int dtype with nulls to integer-na (rather than float) at some point, but currently this is NOT the default. We need some more infrastructure first (see the linked issue).

**Comment by jiangyue12392** [comment]
@jreback Ok, I will wait till the infrastructure is in place after #27283 is sorted out

**Comment by jreback** [comment]
@jiangyue12392 actually part of this PR could partially satisfy that issue

**Comment by jiangyue12392** [comment]
@jreback I suppose the part where the change in `lib.pyx` serves the purpose of distinguishing an `integer_na`. However, on that issue, both @TomAugspurger and @jorisvandenbossche express their concern to keep behaviour consistent with the past where the default datatype is `float64`. I am not sure how users can choose between the new `Int64` behaviour or the old `float64` for backward compatibility

**Comment by jreback** [comment]
so what you do is make the return string integer-na (this is a small api change)
there are a small number of places then that would have to interpret this type (as float) for casting purposes

that’s it

**Comment by jreback** [comment]
@jiangyue12392 can you merge master

**Comment by jiangyue12392** [comment]
> @jiangyue12392 can you merge master

Sorry, I have been busy with other stuff recently, will merge as soon as I get my hands free

**Comment by jreback** [comment]
@jiangyue12392 if you can merge master; I think parts of this patch we would want to merge.

**Comment by jiangyue12392** [comment]
> @jiangyue12392 if you can merge master; I think parts of this patch we would want to merge.

Master merged. Sorry for the delay.

File: pandas/core/frame.py [hunk 1]
```diff
@@ -397,6 +397,7 @@ def __init__(
         columns: Optional[Axes] = None,
         dtype: Optional[Dtype] = None,
         copy: bool = False,
+        to_integer_array: bool = False,
```
Comment (jreback) [inline pandas/core/frame.py]: we can't have this keyword here

File: pandas/_libs/lib.pyx [hunk 2]
```diff
@@ -1951,7 +1953,7 @@ def maybe_convert_numeric(ndarray[object] values, set na_values,
 @cython.wraparound(False)
 def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
                           bint safe=0, bint convert_datetime=0,
-                          bint convert_timedelta=0):
+                          bint convert_timedelta=0, to_integer_array=False):
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: ok with a keyword (here as this is in internal funtion), but can you name
```bint convert_to_nullable_integer=False```

**Review by jreback — CHANGES_REQUESTED** [review]
if you just do the internal change in maybe_convert_objects and add a test for that we can get this in. pls merge master as wel.

File: pandas/core/frame.py (hunk 1 above)
Comment (jiangyue12392) [reply pandas/core/frame.py]: External keywords removed

File: pandas/_libs/lib.pyx (hunk 2 above)
Comment (jiangyue12392) [reply pandas/_libs/lib.pyx]: Variable name refactored.

**Comment by jiangyue12392** [comment]
> if you just do the internal change in maybe_convert_objects and add a test for that we can get this in. pls merge master as wel.

I have made the required changes and added the test.

File: pandas/tests/dtypes/test_inference.py [hunk 3]
```diff
@@ -552,6 +552,15 @@ def test_maybe_convert_objects_datetime(self):
         out = lib.maybe_convert_objects(arr, convert_datetime=1, convert_timedelta=1)
         tm.assert_numpy_array_equal(out, exp)
 
+    def test_maybe_convert_objects_nullable_integer(self):
+        # GH27335
+        arr = np.array([2, np.NaN], dtype=object)
+        result = lib.maybe_convert_objects(arr, convert_to_nullable_integer=1)
+        from pandas.core.arrays import IntegerArray
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py]: can import at the top.

File: pandas/tests/dtypes/test_inference.py [hunk 4]
```diff
@@ -552,6 +552,15 @@ def test_maybe_convert_objects_datetime(self):
...
+
+        exp = IntegerArray(np.array([2, 0], dtype="i8"), np.array([False, True]))
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py]: can you  parameterize this test with an int64 array as well

File: pandas/tests/dtypes/test_inference.py [hunk 5]
```diff
@@ -552,6 +552,15 @@ def test_maybe_convert_objects_datetime(self):
...
+        tm.assert_equal(result, exp)
```
Comment (jreback) [inline pandas/tests/dtypes/test_inference.py]: use tm.assert_extension_array_equal

File: pandas/_libs/lib.pyx [hunk 6]
```diff
@@ -2085,11 +2094,19 @@ def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
 
     if not seen.object_:
         if not safe:
-            if seen.null_:
+            if seen.null_ or seen.nan_:
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: separate issue is we should change line 2083 to use a DatetimeArray (can be separate PR) or here if it works out.

can you also update the doc-string (well add it really :->) thanks for workign on this.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/tests/dtypes/test_inference.py (hunk 3 above)
Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Shifted import statement

File: pandas/tests/dtypes/test_inference.py (hunk 4 above)
Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Parameterized as requested

File: pandas/tests/dtypes/test_inference.py (hunk 5 above)
Comment (jiangyue12392) [reply pandas/tests/dtypes/test_inference.py]: Assertion method changed

File: pandas/_libs/lib.pyx (hunk 6 above)
Comment (jiangyue12392) [reply pandas/_libs/lib.pyx]: I will leave line 2083 out as it is a separate issue.
Which version of the doc-string shall I modify?

File: pandas/_libs/lib.pyx [hunk 7]
```diff
@@ -1955,7 +1957,8 @@ def maybe_convert_numeric(ndarray[object] values, set na_values,
 @cython.wraparound(False)
 def maybe_convert_objects(ndarray[object] objects, bint try_float=0,
                           bint safe=0, bint convert_datetime=0,
-                          bint convert_timedelta=0):
+                          bint convert_timedelta=0,
+                          bint convert_to_nullable_integer=0):
     """
```
Comment (jreback) [inline pandas/_libs/lib.pyx]: if you can update this doc-string here e.g. Returns / Parameters

**Review by jreback — CHANGES_REQUESTED** [review]
small comment, pls rebase and ping on green.

File: pandas/_libs/lib.pyx (hunk 7 above)
Comment (jiangyue12392) [reply pandas/_libs/lib.pyx]: Doc-string added

**Comment by jiangyue12392** [comment]
> small comment, pls rebase and ping on green.

Green now.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jiangyue12392 nice patch. This should enable us to selectively convert to nullable integers with the user facing routines.
